c7n/resources/vpc.py

```diff
@@ -2575,3 +2574,48 @@ def process(self, resources, event=None):
                 results.append(r)
                 r[self.match_annotation_key] = matched
         return results
+
+
+@Subnet.action_registry.register('modify_subnet_attribute')
+class SubnetModifyAtrributes(BaseAction):
+    """Modify subnet attributes.
+
+    :example:
+
+    .. code-block:: yaml
+
+            policies:
+              - name: turn-on-public-ip-protection
+                resource: aws.subnet
+                filters:
+                  - type: subnet
+                    key: "MapPublicIpOnLaunch.enabled"
+                    value: false
+                actions:
+                  - type: modify-subnet-attribute
+                    attributes:
+                      "MapPublicIpOnLaunch: false"
+    """
+
+    schema = type_schema(
+        "modify-subnet-attribute",
+        **{
+            'type': {'enum': ['modify_subnet_attribute']},
+            'map_public_ip_on_launch': {'type': 'boolean'},
+            'assign_ipv6_address_on_creation': {'type': 'boolean'},
+            'map_customer_owned_ip_on_launch': {'type': 'boolean'},
+            'enable_dns_64': {'type': 'boolean'},
+            'enable_resource_name_dns_a_record_on_launch': {'type': 'boolean'},
+            'enable_resource_name_dns_aaaa_record_on_launch': {'type': 'boolean'},
+            'disable_lni_at_device_index': {'type': 'boolean'}
+        })
+
+    permissions = ("ec2:ModifySubentAttributes",)
+
+    def process(self, resources):
+        client = local_session(self.manager.session_factory).client('ec2')
+        map_public_ip_on_launch = self.data.get('map_public_ip_on_launch')
+        for r in resources:
+            client.modify_subnet_attribute(
+                SubnetId=r['SubnetId'], MapPublicIpOnLaunch={'Value': map_public_ip_on_launch})
```

Comment:
the api parameters should be pulled from the policy

Reply:
Yes I am getting it from the policy: 
map_public_ip_on_launch = self.data.get('map_public_ip_on_launch')